File: server/src/main/java/org/hkurh/doky/users/UserFacadeImpl.java
package org.hkurh.doky.users;

import org.hkurh.doky.errorhandling.DokyAuthenticationException;
import org.hkurh.doky.errorhandling.DokyRegistrationException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.lang.NonNull;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Component;

import static java.lang.String.format;
import static org.hkurh.doky.MapperFactory.getModelMapper;

@Component
public class UserFacadeImpl implements UserFacade {

    private static final Logger LOG = LoggerFactory.getLogger(UserFacadeImpl.class);

    private final UserService userService;
    private final PasswordEncoder passwordEncoder;

    public UserFacadeImpl(UserService userService, PasswordEncoder passwordEncoder) {
        this.userService = userService;
        this.passwordEncoder = passwordEncoder;
    }

    @Override
    public void checkCredentials(@NonNull final String userUid, @NonNull final String password) {
        if (!userService.checkUserExistence(userUid)) {
            throw new DokyAuthenticationException("User doesn't exist");
        }
        final UserEntity userEntity = userService.findUserByUid(userUid);
        final String encodedPassword = userEntity.getPassword();
        if (!passwordEncoder.matches(password, encodedPassword)) {
            throw new DokyAuthenticationException("Incorrect credentials");
        }
    }

    @Override
    public UserDto register(@NonNull final String userUid, @NonNull final String password) {
        if (userService.checkUserExistence(userUid)) {
            throw new DokyRegistrationException("User already exists");
        }
        final String encodedPassword = passwordEncoder.encode(password);
        final UserEntity userEntity = userService.create(userUid, encodedPassword);
        LOG.info(format("Register new user [%s]", userEntity));
        return getModelMapper().map(userEntity, UserDto.class);
    }

    @Override
    public UserDto getCurrentUser() {
        final UserEntity userEntity = userService.getCurrentUser();
        return getModelMapper().map(userEntity, UserDto.class);
    }
}

File: server/src/main/java/org/hkurh/doky/users/UserFacadeImpl.kt
package org.hkurh.doky.users

import org.apache.commons.lang3.StringUtils
import org.hkurh.doky.errorhandling.DokyAuthenticationException
import org.hkurh.doky.errorhandling.DokyRegistrationException
import org.hkurh.doky.toDto
import org.slf4j.LoggerFactory
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Component

@Component
class UserFacadeImpl(private val userService: UserService, private val passwordEncoder: PasswordEncoder) : UserFacade {
    override fun checkCredentials(userUid: String, password: String) {
        if (!userService.checkUserExistence(userUid)) throw DokyAuthenticationException("User doesn't exist")
        val userEntity = userService.findUserByUid(userUid)
        val encodedPassword = userEntity!!.password
        if (!passwordEncoder.matches(password, encodedPassword)
        ) throw DokyAuthenticationException("Incorrect credentials")
    }

    override fun register(userUid: String, password: String): UserDto {
        if (userService.checkUserExistence(userUid)) throw DokyRegistrationException("User already exists")
        val encodedPassword = passwordEncoder.encode(password)
        val userEntity = userService.create(userUid, encodedPassword)
        LOG.info("Register new user $userEntity")
        return userEntity.toDto()
    }

    override fun getCurrentUser(): UserDto {
        return userService.getCurrentUser().toDto()
    }

    override fun updateCurrentUser(userDto: UserDto) {
        val currentUser = userService.getCurrentUser()
        if (StringUtils.isNotBlank(userDto.name)) currentUser.name = userDto.name!!
        if (StringUtils.isNotBlank(userDto.password)) currentUser.password = passwordEncoder.encode(userDto.password)
        userService.updateUser(currentUser)
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(UserFacadeImpl::class.java)
    }
}
